README.md
# gfxlang-hs
gfxlang-hs is a transpiler written in Haskell that produces C code that makes use of Tsoding's Olivec library

## Building/Installing
```sh
$ git clone https://github.com/rfmineguy/gfx-lang.git
$ stack install 
```

## Dependencies
This project relies on the presence of Olivec and stbimagewrite
These should be in the same folder as your .gfx file
```sh
$ wget https://raw.githubusercontent.com/tsoding/olive.c/master/olive.c
$ wget https://raw.githubusercontent.com/nothings/stb/master/stb_image_write.h
```

## Usage
I only support producing a png file at the moment (and probably forever) so keep that in mind
```sh
$ gfxlang-hs-exe -f a.gfx -o a.c
$ gcc a.c -o a
$ ./a
```

NOTE: Without the Dependencies "a.c" will not be able to compile

## Examples
```
settings () {
  outFile = "image.png"
  width = 600
  height = 600
}
run () {
  fill(0xFF454545)
  circle(300 300 180 0xFF2D00BC)
  circle(300 300 90 0xFF459832)
  rect(30 30 50 50 0xFF994213)
  line(20 40 400 300 0xFF532156)
}
```
![Example](test/image.png)

## Explaination
- settings () is a required function where you specify all of the varaibles you will use
    - Currently the only variables used here are required
        - outFile -> Specifies the output filename (will always be png format)
        - width   -> Specifies the width of the output image
        - height  -> Specifies the height of the output image
- run () is another required function you put all the code that draws stuff
    - The functions callable in here are the same as those present in the normal OliveC library which can be found here [OliveC](https://github.com/tsoding/olive.c/tree/master)




CHANGELOG.md
# Changelog for `gfxlang-hs`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/a.c
#define OLIVEC_IMPLEMENTATION
#include "olive.c"
#include <stdlib.h>
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
Olivec_Canvas oc;
uint32_t *pixels;
const char* outFile;
int height;
int width;
void settings(){
	outFile = "image.png";
	width = 600;
	height = 600;
}
void drawPlayer(){
}
void run(){
	olivec_fill(oc,4282729797);
	olivec_circle(oc,300,300,180,4281139388);
	olivec_circle(oc,300,300,90,4282751026);
	olivec_rect(oc,30,30,50,50,4288234003);
	olivec_line(oc,20,40,400,300,4283638102);
}
int main(void) {
	settings();
	pixels = malloc(width * height * sizeof(uint32_t));
	oc = olivec_canvas(pixels, width, height, width);
	run();
	if (!stbi_write_png(outFile, width, height, 4, pixels, sizeof(uint32_t)*width)) { return 1; }
	free(pixels);
	return 0;
 }

app/Main.hs
module Main (main) where

{-
 Example source
  
  settings {
    size 600 600
    framerate 30
  }
  begin {
    x = 4
    y = 6
    angle = 0
  }
  draw {
    angle += 3
  }

============================

void settings() {
  size(600, 600)
}

int main () {
  
}

-}


import Text.Parsec as P
import Text.Parsec.String as P
import Text.ParserCombinators.Parsec.Number as P
import Options.Applicative as O
import System.IO
import Data.Set

--- AST Decl
type Var = String
data Param = IntParam Int | StringParam String
  deriving (Eq, Show)
data Statement = Assign Var Int | Call String [Param] 
  deriving (Eq, Show)
data Procedure = Section String [Param] [Statement] 
  deriving (Eq, Show)
type Code = [Procedure]

--
--- Parsing
--
assignParser :: P.Parser Statement
assignParser = Assign <$> many1 letter <*> (spaces >> char '=' >> spaces >> int)

callParser :: P.Parser Statement
callParser = Call <$> many1 letter <*> multiParamParser

statementParser :: P.Parser Statement
statementParser = P.try assignParser P.<|> P.try callParser

multiStatementParser :: P.Parser [Statement]
multiStatementParser = P.spaces >> P.many (statementParser <* P.spaces)

paramParser :: P.Parser Param
paramParser = ((IntParam <$> int) P.<|> (StringParam <$> many1 letter)) <* spaces

multiParamParser :: P.Parser [Param]
multiParamParser = spaces >> char '(' >> P.many (spaces >> paramParser) <* spaces <* char ')'

blockParser :: P.Parser [Statement]
blockParser = spaces >> char '{' >> multiStatementParser <* char '}'

procedureParser :: P.Parser Procedure
procedureParser = Section <$> (spaces >> many1 letter) <*> (P.many space >> multiParamParser) <*> (P.many space >> blockParser) <* spaces

codeParser :: P.Parser Code
codeParser = P.many procedureParser <* eof

--
--- Extraction
--
extractVariablesProc :: Procedure -> [Var]
extractVariablesProc (Section s ps stmts) = extractVariablesStatements stmts

extractVariablesStatements :: [Statement] -> [Var]
extractVariablesStatements = concatMap extractVariablesStatement

extractVariablesStatement :: Statement -> [Var]
extractVariablesStatement (Assign var val) = [var]
extractVariablesStatement (Call s ps) = []

--
--- Codegen
--
codeGenerator :: Code -> String
codeGenerator code = boilerplateHead ++ varDeclsGenerator vars ++ concatMap procedureGenerator code ++ boilerplateTail
  where vars = toList (fromList (concatMap extractVariablesProc code))

boilerplateHead :: String
boilerplateHead = "void size(int, int);\n" ++ "void settings();\n" ++ "void begin();\n" ++ "void draw(float dt);\n"

boilerplateTail :: String
boilerplateTail = "#include \"gfxcore.c\"\n"

procedureGenerator :: Procedure -> String
procedureGenerator (Section s p stmts) = "void " ++ s ++ "(" ++ paramsGenerator p ++ ")" ++ "{\n" ++ statementsGenerator stmts ++ "}\n"

paramsGenerator :: [Param] -> String
paramsGenerator [] = ""
paramsGenerator [p] = paramGenerator p
paramsGenerator (p:ps) = paramGenerator p ++ "," ++ paramsGenerator ps

paramGenerator :: Param -> String
paramGenerator (StringParam i) = "int " ++ i
paramGenerator (IntParam _) = undefined

statementsGenerator :: [Statement] -> String
statementsGenerator = concatMap statementGenerator

statementGenerator :: Statement -> String
statementGenerator (Assign v a) = "\t" ++ v ++ " = " ++ show a ++ ";\n"
statementGenerator (Call s ps) = "\t" ++ s ++ "(" ++ argsGenerator ps ++ ");\n"

argsGenerator :: [Param] -> String
argsGenerator [] = ""
argsGenerator [p] = argGenerator p
argsGenerator (p:ps) = argGenerator p ++ "," ++ argsGenerator ps

argGenerator :: Param -> String
argGenerator (IntParam i) = show i
argGenerator (StringParam s) = s

varDeclsGenerator :: [Var] -> String
varDeclsGenerator = concatMap varDeclGenerator

varDeclGenerator :: Var -> String
varDeclGenerator v = "int " ++ v ++ ";\n"

--
--- Arg parsing
--
data Options = Options
  { file :: String,
    output :: String
  }

options :: O.Parser Options
options = Options
        <$> O.strOption ( long "file" <> short 'f' <> help "Supply file" )
        <*> O.strOption ( long "output" <> short 'o' <> help "Output file" ) 

opts :: ParserInfo Options
opts = info (options <**> helper)
        ( fullDesc
        <> progDesc "Transpile gfx program to C"
        <> header "" )

main :: IO ()
main = run =<< execParser opts

run :: Options -> IO ()
run (Options s o) = do
  putStrLn $ "{InputFile: " ++ s ++ " ; OutputFile: " ++ o ++ "}"
  handle <- openFile s ReadMode
  contents <- hGetContents handle
  let c = parse codeParser "" contents
  case c of
    Left e -> print e
    Right code -> do {
      let variables = toList (fromList (concatMap extractVariablesProc code))
    ; let decl = varDeclsGenerator variables
    ; print ("Variables = ", variables, " => {" ++ decl ++ "}")
    ; writeFile o (codeGenerator code)
    }
  print c
